src/stan/variational/advi.hpp

```diff
@@ -528,22 +531,26 @@ namespace stan {
            << n_posterior_samples_
            << " from the approximate posterior... ";
         logger.info(ss);
-
+        double log_p = 0;
+        double log_q = 0;
+        // Draw posterior samples. log_q is the log normal densities.
         for (int n = 0; n < n_posterior_samples_; ++n) {
-          variational.sample(rng_, cont_params_);
+          variational.sample_lp(rng_, cont_params_, log_q);
           for (int i = 0; i < cont_params_.size(); ++i) {
             cont_vector.at(i) = cont_params_(i);
           }
           std::stringstream msg2;
-          model_.write_array(rng_, cont_vector, disc_vector, values,
-                             true, true, &msg2);
+          model_.write_array(rng_, cont_vector, disc_vector,
+                             values, true, true, &msg2);
+          //  log_p: Log probability in the unconstrained space
```

Comment:
No, the constants are not necessary for lp__. Indeed they should not even been included for elbo when doing optimization. However it seems everything accidentally becomes 0 when I tried to drop those constants (https://discourse.mc-stan.org/t/dropping-the-constant-in-log-prob/3889/6)

Reply:
Let's fix this first. Is this true for any model?

To reproduce this, all I need to do is make a model, print line 546 with propto = true, and I'll get zeros?

Reply:
Oh, well that's resolved I guess: https://github.com/stan-dev/math/blob/develop/stan/math/prim/scal/prob/normal_lpdf.hpp#L65

If all the template arguments are double, the log densities just short circuit and returns 0.0. I guess the way this is all coded, any double inputs are interpreted as not model parameters. So this looks like a design decision -- no bug to fix.

Probly what you have is right enough. You could pass in the parameters as vars if you really needed to get the proportionality stuff right, but then you'd have to be careful to clean up the autodiff stack again.

Reply:
I made some tests again.  It is not zero, but if I replace line 546 with 
`log_p =model_.template log_prob<false, true>(cont_params_, &msg2) - model_.template log_prob<true, true>(cont_params_, &msg2);`
I can find it is not a constant.
As I wrote in discourse, log_prob<true, true> does return the correct density with constants. But log_prob<false, true> is wrong.

Reply:
Can you try:

```
log_p =model_.template log_prob<false, true>(stan::math::to_var(cont_params_), &msg2) - model_.template log_prob<true, true>(stan::math::to_var(cont_params_), &msg2);
```

real quick? That'll convert all the doubles to vars, which is implementation that would be used by all the inference algorithms.

Reply:
Actually sorry, you'll need another value_of:

```
log_p = value_of(model_.template log_prob<false, true>(stan::math::to_var(cont_params_), &msg2) - model_.template log_prob<true, true>(stan::math::to_var(cont_params_), &msg2));
```

So the output is a scalar and not a var.

Reply:
It leads to many other errors. 
``error: no matching member function  for call to 'log_prob'``

Reply:
Anyway I guess it will be better to drop the proportionality constants. But the additional computational cost for keeping them is also ignorable.

Reply:
Yuck, can you point me at a model and data where this is happening? I've got some super super hacky code to test stuff like this but it's hard to explain how to use it.

Reply:
Yeah, I agree, it's not worth worrying about the computations. We should just nail down whatever is going on here so if it comes up again we aren't surprised.

re: previous remark -- is it just any model?

Reply:
I think it is any model.

Reply:
@yao-yl thanks for pointing this out. Just use propto = false in this pull. I dug around a little and made a simple test for this so it can be documented somewhere: https://github.com/stan-dev/math/issues/1020